drivers/sensor/akm09918c/akm09918c.c

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright (c) 2023 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT asahi_kasei_akm09918c
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/logging/log.h>
+
+#include "akm09918c.h"
+#include "akm09918c_reg.h"
+
+LOG_MODULE_REGISTER(AKM09918C, CONFIG_SENSOR_LOG_LEVEL);
+
+static int akm09918c_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	struct akm09918c_data *data = dev->data;
+	const struct akm09918c_config *cfg = dev->config;
+	uint8_t buf[9] = {0};
+
+	if (chan != SENSOR_CHAN_ALL && chan != SENSOR_CHAN_MAGN_X && chan != SENSOR_CHAN_MAGN_Y &&
+	    chan != SENSOR_CHAN_MAGN_Z && chan != SENSOR_CHAN_MAGN_XYZ) {
+		LOG_WRN("Invalid channel %d", chan);
+		return -EINVAL;
+	}
+
+	if (data->mode == AKM09918C_CNTL2_PWR_DOWN) {
+		if (i2c_reg_write_byte_dt(&cfg->i2c, AKM09918C_REG_CNTL2,
+					  AKM09918C_CNTL2_SINGLE_MEASURE) != 0) {
+			LOG_ERR("Failed to start measurement.");
+			return -EIO;
+		}
+
+		/* Wait for sample */
+		LOG_DBG("Waiting for sample...");
+		k_busy_wait(AKM09918C_MEASURE_TIME_US);
+	}
+
+	/* We have to read through the TMPS register or the data_ready bit won't clear */
+	if (i2c_burst_read_dt(&cfg->i2c, AKM09918C_REG_ST1, buf, ARRAY_SIZE(buf)) != 0) {
+		LOG_ERR("Failed to read sample data.");
+		return -EIO;
+	}
+
+	if (FIELD_GET(AKM09918C_ST1_DRDY, buf[0]) == 0) {
+		LOG_ERR("Data not ready, st1=0x%02x", buf[0]);
+		return -EBUSY;
+	}
+
+	data->x_sample = sys_le16_to_cpu(buf[1] | (buf[2] << 8));
+	data->y_sample = sys_le16_to_cpu(buf[3] | (buf[4] << 8));
+	data->z_sample = sys_le16_to_cpu(buf[5] | (buf[6] << 8));
+
+	return 0;
+}
+
+static void akm09918c_convert(struct sensor_value *val, int16_t sample)
+{
+	int64_t conv_val = sample * AKM09918C_MICRO_GAUSS_PER_BIT;
+
+	val->val1 = conv_val / 1000000;
+	val->val2 = conv_val % 1000000;
```

Comment:
Do you really need a modulo here? Multiply and subtract, conv_val - val1*1000000 and you have the remainder.

Reply:
Done